Web User API Bulk user data lookup and filter via ES

Jtang-1 opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->
This PR improves the performance for getting list of web users via web user API. It also adds back the changes that were removed in this revert [PR](https://github.com/dimagi/commcare-hq/pull/35897): remove `tableau groups` from GET list response for performance reasons. It currently makes a request for each user in the response leading to high latency.

## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->
[USH-5529](https://dimagi.atlassian.net/browse/USH-5529)
With this change, web user API does user data lookup in bulk. This is done the same way as mobile worker API via `UserQuerySetAdapter`

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
No FF

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Tested locally and it leverages an existing class that mobile worker API uses.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
There exists `TestWebUserResource` that tests the `GET` request returns expected users in its response. It also tests that filters correctly filters out users.

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
no QA

### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
<!-- https://commcare-hq.readthedocs.io/migrations_in_practice.html -->
There is no migration in this PR. This PR originally added a filter for `last_modified` but that has since been removed. To support that filter, a migration was done in a previous [PR]
(https://github.com/dimagi/commcare-hq/pull/35824).

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change


[USH-5529]: https://dimagi.atlassian.net/browse/USH-5529?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Jtang-1 committed:
feat: use UserQuerySet to get web users via elasticsearch and to prime user data

Jtang-1 committed:
feat: filter via ES instead of in memory

Jtang-1 committed:
test: populate ES index for test

MartinRiese commented on corehq/apps/api/resources/v0_1.py:
```diff
@@ -176,26 +176,23 @@ def obj_get_list(self, bundle, **kwargs):
-        return results
+            filters.append(SIMPLE_FILTERS['username'](username))
+        queryset = UserQuerySetAdapter(domain, show_archived=False, is_web_user=True, filters=filters)
+        return queryset
```
Is this changing the return value from results to query set? I don't see any changes to the caller code.

Jtang-1 replied:
Yes it changes the results to the queryset. `obj_get_list` is called internally by Tastypie and it expects the return to be "a list-like object of ``Resources``. This is typically a ``QuerySet`` but can be anything that implements slicing. Required.`

[This](https://github.com/django-tastypie/django-tastypie/blob/bd07c587c86c03592dde35910585f2212719fc6d/tastypie/resources.py#L1346-L1349) is where obj_get_list is called and the [paginator](https://github.com/django-tastypie/django-tastypie/blob/bd07c587c86c03592dde35910585f2212719fc6d/tastypie/paginator.py#L31-L33) class it ends up being passed to.

Since `UserQuerySetAdapter` overrides `__getitem__` to implement slicing, it's a valid object to return/

esoergel reviewed (approved):
Are there any sneaky changes to the API output from the switch from couch to ES?

Jtang-1 committed:
avoid dehydrating tableau groups when bulk GET web users.

The tableau group for the user requires a request to tableau. This is a large performance
hit when this needs to be done for a large number of web users. This is a currently a limitation
until we implement a way to bulk get tableau groups for users.

Jtang-1 commented:
@esoergel Not that I see from testing and not that I would expect either since ES only effects the query. The API output should only be affected by how Tastypie processes the CouchUser object that's fetched from the query. And the processing isn't changed in this PR.

Jtang-1 committed:
test: clear cache to not affect followup tests

Jtang-1 commented:
closing and reopening to run tests

Jtang-1 committed:
Merge branch 'master' into jt/web-user-api-bulk-user-data-lookup

Jtang-1 committed:
test: last_modified is not added to commcareuser response

Jtang-1 committed:
test: last_modified is not added to commcareuser response

kaapstorm commented on corehq/apps/api/resources/v0_1.py:
```diff
@@ -159,12 +161,24 @@ class Meta(UserResource.Meta):
         username = bundle.request.GET.get('web_username')
+
+        SIMPLE_FILTERS = {
+            'username': user_es.username
+        }
+        COMPOUND_FILTERS = {
+            'last_modified': make_date_filter(user_es.last_modified)
+        }
+
+        filters = []
+        for key in bundle.request.GET.keys():
+            if '.' in key and key.split(".")[0] in COMPOUND_FILTERS:
+                prefix, qualifier = key.split(".", maxsplit=1)
+                filters.append(COMPOUND_FILTERS[prefix](qualifier, bundle.request.GET.get(key)))
+
         if username:
-            user = WebUser.get_by_username(username)
-            if not (user and user.is_member_of(domain) and user.is_active):
-                user = None
-            return [user] if user else []
-        return list(WebUser.by_domain(domain))
+            filters.append(SIMPLE_FILTERS['username'](username))
+        queryset = UserQuerySetAdapter(domain, show_archived=False, is_web_user=True, filters=filters)
+        return queryset
```
It's not clear to me what the dictionaries are for. I would understand if there were many prefixes. Will we be adding more prefixes?

If there is just one prefix, something like this might be more readable:

```python
    def obj_get_list(self, bundle, domain, **kwargs):
        filters = []
        for key, value in bundle.request.GET.items():
            if '.' in key:
                prefix, qualifier = key.split(".", maxsplit=1)
                if prefix == 'last_modified':
                    last_modified_filter = make_date_filter(user_es.last_modified)
                    filters.append(last_modified_filter(qualifier, value))
        if username := bundle.request.GET.get('web_username'):
            filters.append(user_es.username(username))
        return UserQuerySetAdapter(
            domain,
            show_archived=False,
            is_web_user=True,
            filters=filters,
        )
```

Jtang-1 replied:
I agree that your suggestion is more readable. The dictionaries are a pattern we use in other APIs and not as of now we won't be adding more prefixes, but I think it's a likely that more will be added in the future. So given that, I would prefer to keep the dictionary.

Jtang-1 committed:
remove last_modified from api

This filter may not be neccessary. The original intention was to avoid getting all web users. However,
this ran into an issue where getting just the last modified users will not signal when a user has been
removed from the domain.

Jtang-1 commented:
bump for review

MartinRiese reviewed (approved):
Commits since Ethans last approval look fine
